Discourse integration

JKarlavige opened the pull request:
## Description & motivation

Asana task: https://app.asana.com/0/1200099998847559/1202844016314445/f

This adds a new `DiscourseFeed` component to be used throughout the site. This component uses the Discourse API to display topics on docs pages. A new `/forum` page is created to display different sections of Discourse topics via the new component.

This also includes an example of how the component can be used on any docs pages.

I will be doing the website work within this PR, and will open a separate PR to handle the Algolia search side of this task. I expect the Algolia work to take longer so no reason to hold up this work.

## Preview
Forum page: https://deploy-preview-2004--docs-getdbt-com.netlify.app/forum

## TODOs
- [x] Remove example use of Discourse component from Jinja docs page

JKarlavige committed:
initial component created

JKarlavige committed:
add title

JKarlavige committed:
add netlify function to get discourse posts

JKarlavige committed:
add styling with demo content

JKarlavige committed:
connect page to function

JKarlavige committed:
convert to topics endpoint

JKarlavige committed:
setup props for discourse component

JKarlavige committed:
Merge branch 'current' into discourse-integration

JKarlavige committed:
connect component to search discourse endpoint

JKarlavige committed:
adjust styles

JKarlavige committed:
set comments

JKarlavige committed:
adjust section padding

JKarlavige committed:
encode URI component

JKarlavige committed:
remove console

JKarlavige committed:
adjust section styles

JKarlavige committed:
set default minHeight and allow custom styles

JKarlavige committed:
spread styles

JKarlavige committed:
set initial min-height

joellabes commented on website/docs/faqs/Jinja/which-jinja-docs.md:
```diff
@@ -10,3 +10,6 @@ If you are stuck with a Jinja issue, it can get confusing where to check for mor
 3. [Agate's table docs](https://agate.readthedocs.io/page/api/table.html): If you're operating on the result of a query, dbt will pass it back to you as an agate table. This means that the methods you call on the <Term id="table" /> belong to the Agate library rather than Jinja or dbt.
+
+<!-- EXAMPLE: DELETE BEFORE TAKE-LIVE -->
+<DiscourseFeed title='Discourse Jinja topics' tags='jinja' styles={{marginTop: 40}} />
```
Could the top margin be baked into the default template? I suspect that it's something we'll always want.

joellabes commented on website/functions/get-discourse-posts.js:
```diff
@@ -0,0 +1,100 @@
+  } catch(err) {
+    // Log and return the error
+    console.log('err', err)
+    return await returnResponse(500, { error: 'Unable to get events from Discourse.'})
```
I assume this is copy-paste from the events feed
```suggestion
    return await returnResponse(500, { error: 'Unable to get topics from Discourse.'})
```

joellabes commented on website/functions/get-discourse-posts.js:
```diff
@@ -0,0 +1,100 @@
+function validateItem({ key, value }) {
+  // predefined Discourse values
+  // https://docs.discourse.org/#tag/Search/operation/search
+  const inStringValues = ['title', 'likes', 'personal', 'messages', 'seen', 'unseen', 'posted', 'created, watching', 'tracking', 'bookmarks', 'assigned', 'unassigned', 'first', 'pinned', 'wiki']
```
```suggestion
  const inStringValues = ['title', 'likes', 'personal', 'messages', 'seen', 'unseen', 'posted', 'created', 'watching', 'tracking', 'bookmarks', 'assigned', 'unassigned', 'first', 'pinned', 'wiki']
```

joellabes replied:
I don't know which user this API key is associated with, but it's probably unnecessary to support things like personal, seen, unseen, created as they're user-specific. I guess they're not doing any harm? 

If we wanted to cut it down, I think the only ones that are generic are `title, first, pinned, wiki`.

JKarlavige replied:
Yeah it doesn't hurt having those on our end as it only ensures the values will work with Discourse. But if you would like to prevent those other values from being used, we can remove them from the validation here. Will leave that up to you!

joellabes replied:
I think let's remove them so that it's clear to future users which ones they can realistically use

joellabes commented on website/functions/get-discourse-posts.js:
```diff
@@ -0,0 +1,100 @@
+      } else if(key === 'inString') {
+        query += `in:${value}`
+      } else {
+        query += `${key}:${value} `
```
Does MDX allow duplicate keys? You're allowed to do something like [`status:open status:solved`](https://discourse.getdbt.com/search?q=status%3Aopen%20status%3Asolved), but this feels like it would break: 

`<DiscourseFeed category='help' status='solved' status='open' />`

It's not likely that we'd do this very often, so having a string instead of an array as the main input type is more ergonomic. Can you pass arrays into MDX if needed (`status=['solved', 'open']`)? Or would we have to comma-separate them and you split them out?

JKarlavige replied:
Yeah that's a good point, i'll adjust this to support multiple values for the same property.

JKarlavige replied:
Updated this to support arrays for the `status` property. The docs page example shows how multiple statuses can be passed in: https://github.com/dbt-labs/docs.getdbt.com/blob/2590c0049fcdd15eb0005f10cd4c625f799d2528/website/docs/faqs/Jinja/which-jinja-docs.md

`status` can now support string: `status='open'` or arrays: `status={['open', 'public']}` when being passed into the component

joellabes commented on website/src/components/discourse/index.js:
```diff
@@ -0,0 +1,122 @@
+      {loading ? (
+        <img src="/img/loader-icon.svg" alt="Loading" className={feedStyles.loadingIcon} />
+      ) : isError || !posts?.length > 0 ? (
+        <p>Unable to load Discourse posts at this time.</p>
```
```suggestion
        <p>Unable to load forum posts at this time.</p>
```

I think we'll de-emphasise the platform we're using and lean into it just being the Forum (in the UI at least; the code can still be specific)

joellabes commented on website/src/components/discourse/index.js:
```diff
@@ -0,0 +1,122 @@
+  term = undefined,
+  category = undefined,
+  title = undefined,
+  link_text = "See latest topics",
+  link_href = "https://discourse.getdbt.com/",
```
As the default, it would be good for the CTA to be `Ask the Community` which goes to `https://discourse.getdbt.com/new-topic?category=XXXX&tags=comma,separated`

Note that in search you can either use `,` in tags (for any of) and `+` (for all of), so you'd need to replace +s with ,s here too

https://meta.discourse.org/t/how-to-create-a-new-topic-link-with-pre-filled-information/28074

Then we could override it on the forum homepage to be the Read More ones you currently have. (Side note: are we allowed to set them to null/undefined to cause no button to be shown?)

JKarlavige replied:
Updated the default text & link for the CTA. If the `category` or `tags` are passed in, it will append them to the link url by default. If a custom `link_href` is passed in, it will override the default.

For the tags, since it's a string with comma (or +) separated values, passing a string such as `<DiscourseFeed  tags='jinja,testing' />` will work as i'm passing the tags string directly into the query.

Also added a separate `hide_cta` prop to control whether the cta should be hidden.

joellabes commented on website/src/components/discourse/index.js:
```diff
@@ -0,0 +1,122 @@
+export default function DiscourseFeed({
+  order = 'latest_topic',
+  status = undefined,
+  after = undefined,
```
Can we have relative periods here? I was thinking we'd do something like last 30 days for the help ones, and last 90 for discussions.

JKarlavige replied:
Does this refer to the `after` property? If a different category is passed in other than `help` or `discussions` should we exclude the `after` date, or should we set a default number of days back?

I pushed an update to set a relative `after` date if the category is `help` or `discussions`, otherwise it excludes the `after` date altogether for now.

joellabes replied:
Yes I meant it for after - will have a look at the new code

joellabes replied:
I had originally thought it might be another prop on the DiscourseFeed, but I don't mind how you've done it here! 

Everything that isn't `help` can be 90 days - we might wind up reconsidering this at some point but messing around with the params is the sort of thing I can change unassisted 🎉

joellabes commented on website/src/components/discourse/index.js:
```diff
@@ -0,0 +1,122 @@
+        <ul>
+          {posts.map(post => (
+            <li>
+              <PostWrapper post={post}>{post.title}</PostWrapper>
```
As they currently look, there's not a lot of social proof that the answers you click through to are going to be accurate. Would it be overkill to put a ✅ with `Solved` hover text in the front of solved topics? (I think it's called `has_accepted_answer` or something in the response body)

JKarlavige replied:
Love this

joellabes commented on website/src/components/discourse/index.js:
```diff
@@ -0,0 +1,122 @@
+                <>
+                  {' '}-
+                  <span>
+                    {post?.username && `by ${post.username}${post?.posts_count && ','}`}
```
If you wanted to get the username without a second API call, you can get it from the `posts` part of the search response. 
![image](https://user-images.githubusercontent.com/7335046/190550272-ca44eec0-3078-4453-a772-11e6a15bac9d.png)

You'll want to only take it if the `post_number` is 1. When we're searching for topic-wide metadata (e.g. tag, category, solved status) this is fine - that'll always be the one you're given. It's only if you search for specific text that you might get a different post - e.g. this one, where you get the first post in a topic that contains the search text. We don't intend to do that for any of our embeds AFAIK. 
![image](https://user-images.githubusercontent.com/7335046/190550622-0b0890b0-a46f-40c4-aa17-628573b36d5a.png)

JKarlavige replied:
ahhhh good find! Will get this in there

joellabes commented on website/src/components/discourse/index.js:
```diff
@@ -0,0 +1,122 @@
+                  <span>
+                    {post?.username && `by ${post.username}${post?.posts_count && ','}`}
+                    {' '}
+                    {post?.posts_count && `${post.posts_count} comments`}
```
Arguably this could be `{posts_count - 1}` replies. Not attached to this. 

Side note: `1 comments` is my biggest UI pet peeve. (Very attached to this 😅)

joellabes reviewed (changes requested):
This is looking so great @JKarlavige 🤩😍

A handful of comments, hopefully nothing too arduous!

joellabes commented on website/src/components/discourse/index.js:
```diff
@@ -0,0 +1,122 @@
+  max_views = undefined,
+  tags = undefined, 
+  term = undefined,
+  category = undefined,
```
Can we default this to `'help'` so we don't have to specify it everywhere? That's going to be the overwhelming majority of embeds.

joellabes commented on website/src/components/discourse/index.js:
```diff
@@ -0,0 +1,122 @@
+
+export default function DiscourseFeed({
+  order = 'latest_topic',
+  status = undefined,
```
Likewise here, let's have status:solved as the default and override it where needed. 

OR, it might be easier to have a focused `DiscourseHelpFeed` which ultimately creates a `DiscourseFeed` but we don't have to put defaults into the main constructor. Is that practical?

JKarlavige replied:
Set the default status to `solved`. We can create a more specific component which uses the original component with specific props if you'd like. If you feel like it would be easier for the team to work with we can get it setup. I would just want to go over all the settings that more specific component would need.

joellabes replied:
Yeah I think it would be good to make a focused one. The default props: 
```javascript
order = 'latest_topic',
status = 'solved',
category = 'help',
link_text = 'Ask the Community',
link_href = // the same URL builder you've already got and I don't want to copy-paste
hide_cta = 'false'
```

Then on the main `DiscourseFeed`, everything would be undefined except for `hide_cta` which would be true. (Or, if you [changed it to be `show_cta`](https://github.com/dbt-labs/docs.getdbt.com/pull/2004#discussion_r974851260), then _everything_ could be undefined because `undefined` is falsy right?)

joellabes reviewed (commented):
Was just writing a comment in Asana about how we only have to specify the tags we want in help mode, and then thought "hang on a minute, is that accurate?"

JKarlavige committed:
add default top margin

JKarlavige committed:
Update website/functions/get-discourse-posts.js

Co-authored-by: Joel Labes <joel.labes@dbtlabs.com>

JKarlavige committed:
Update website/functions/get-discourse-posts.js

Co-authored-by: Joel Labes <joel.labes@dbtlabs.com>

JKarlavige committed:
Update website/src/components/discourse/index.js

Co-authored-by: Joel Labes <joel.labes@dbtlabs.com>

JKarlavige committed:
set default cta link and add ability to hide cta

JKarlavige committed:
add checkbox for solved posts

JKarlavige committed:
set relative after date for help and discussions categories

JKarlavige committed:
Merge branch 'discourse-integration' of https://github.com/dbt-labs/docs.getdbt.com into discourse-integration

JKarlavige committed:
set default status and category

JKarlavige committed:
remove console

JKarlavige committed:
support array for status property

JKarlavige committed:
get topic author

JKarlavige committed:
change 'post' to 'topic' throughout component and endpoint

JKarlavige committed:
display author next to post link

JKarlavige committed:
adds meta fields

JKarlavige committed:
change comments to replies and adjust for correct reply count

JKarlavige commented:
@joellabes marked comments as resolved for which i pushed updates for. I have a few open questions/clarifications for the others!

JKarlavige committed:
remove console

lfracalossiFA commented:
Hi @JKarlavige,
Is the "-" between the thread title and the byline by design? I don't love it. But I also don't know if this is something that was requested!

<img width="981" alt="Screen Shot 2022-09-19 at 4 40 00 PM" src="https://user-images.githubusercontent.com/72401230/191112380-42a26275-5a34-4c14-a5da-6859ec640113.png">

lfracalossiFA commented:
The fonts also seem small to me on the list items on /forum. (I think it's the overall docs font face; I feel like it skews small). Any objection to bumping those up to 18px? I'd also suggest making the title bold, eg:

<img width="890" alt="Screen Shot 2022-09-19 at 4 47 38 PM" src="https://user-images.githubusercontent.com/72401230/191113562-c437659f-3f69-4089-bda1-9d5f8a229622.png">

lfracalossiFA commented on website/src/components/discourse/styles.module.css:
```diff
@@ -0,0 +1,45 @@
+  padding: 0;
+  width: 100%;
+}
+.discourseTopics ul li {
+  padding: 1rem 0;
+  border-bottom: 2px solid var(--color-off-white);
+}
```
I mentioned this in the discussion section of this PR, but I think we should bump up the font here to 18px, and make the `a` bold. 

(Maybe just for /forum? Not sure it would be appropriate in the in-page component, because it might distract from the doc content itself).

JKarlavige replied:
I bumped the font to `18px` and make it bold. I'm not crazy about having bold text in one place where the component is used, and not bold in another spot. We'd have to add another prop to control that, and there's already 18 potential props that can be used. 😅 Maybe `font-weight: 600` is a good in-between? I'll have that in my next push

lfracalossiFA replied:
Works for me, thanks!

lfracalossiFA commented on website/src/components/discourse/styles.module.css:
```diff
@@ -0,0 +1,45 @@
+}
+
+/* Cta */
+.discourseCta {
+  font-size: 1.1rem;
+  margin-top: auto;
+}
```
I feel like the font size and weight are inconsistent with buttons like this elsewhere on the site? (Compare to homepage). But maybe there is a different button you based this off?

JKarlavige replied:
Hmmmm, this button is the same one used on the homepage cards, i'm using the same `button button--primary` classes. The only thing I did was reduce the font-size a bit as i felt it was too large. Is there anything specific you think we should change with the buttons for this component?

lfracalossiFA replied:
> Is there anything specific you think we should change with the buttons for this component?

Hm, I guess not. From a UX perspective it's just as usable.

lfracalossiFA commented:
It looks like the meta description is "Description here." Any way to customize that? 

<img width="648" alt="Screen Shot 2022-09-19 at 5 18 22 PM" src="https://user-images.githubusercontent.com/72401230/191118502-d06cec45-7c52-48a1-8c5e-1601aab92326.png">

lfracalossiFA commented on website/src/components/discourse/index.js:
```diff
@@ -0,0 +1,147 @@
+    fetchData()
+  }, [])
+
+  // Set initial min-height
+  // This is to avoid layout shifts
+  // which affects Lighthouse performance scores
+  const setMinHeight = isError
+    ? 'auto'
+    : 414
```
👍

lfracalossiFA commented on website/src/pages/forum.js:
```diff
@@ -0,0 +1,25 @@
+        <div className='container'>   
+          <h1>dbt Discourse</h1>
+          <DiscourseFeed title='Open topics' category='help' status='unsolved'  link_text='See open topics' />
+          <DiscourseFeed title='Cool Problem Solving' category='show-and-tell' after='2022-06-14' status='unsolved' link_text='See more topics' link_href='https://discourse.getdbt.com/c/show-and-tell/22' />
```
I saw the conversation about the `after` attribute with Joel above -- will we want a date like this to be hard-coded here? Or is this just an example?

joellabes replied:
I doubt we'd ever hardcode a date in here, and instead would just use the relative calculation 👍

lfracalossiFA replied:
@joellabes That's why I'm wondering about the `after='2022-06-14'` on Cool Problem Solving...

JKarlavige replied:
Removed the `after` property for the Cool Problem Solving component.

lfracalossiFA reviewed (dismissed):
Overall, this is really looking good -- will add great value to our pages. I have a few questions as well as requests re: styling/UI. (I didn't see a comp anywhere, so feel free to ignore my suggestions if they contradict any stakeholder requests).

joellabes commented on website/src/components/discourse/index.js:
```diff
@@ -34,41 +84,64 @@ export default function DiscourseFeed(styles = {}) {
+              {topic?.has_accepted_answer && (
+                <span className={feedStyles.solvedTopic} title="Solved">✅ </span>
+              )}
+              <TopicWrapper topic={topic}>{topic.title}</TopicWrapper>
```
Also on the social proof front, it'd be nice to pull through the `like_count` from the posts array. I think this will have more purpose on the Discussion ones as I doubt people are going to click like on the OP of a support request, but if they do we can show them there too! 

I don't know how you render them - it might get messy to have `How do I do computers? by joellabes, 1 reply, 3 likes` but if we don't have design time then that might just have to do for now.

JKarlavige replied:
Here's an example of how it looks with the like count added. Not many have the like count so not sure if any would show up in the preview, so adding it here. I had to set the `after` prop to `2022-01-01` on the first component within the /forum page to get it to show up.

`<DiscourseFeed title='Open topics' category='help' status='unsolved'  link_text='See open topics' link_href='https://discourse.getdbt.com/c/help/19' show_cta={true} after='2022-01-01' />`

![image](https://user-images.githubusercontent.com/13647110/191303169-18d2ffdb-436e-452d-b7a6-5183712e66fc.png)

joellabes replied:
Works for me!

joellabes commented on website/src/pages/forum.js:
```diff
@@ -0,0 +1,25 @@
+          <h1>dbt Discourse</h1>
+          <DiscourseFeed title='Open topics' category='help' status='unsolved'  link_text='See open topics' />
+          <DiscourseFeed title='Cool Problem Solving' category='show-and-tell' after='2022-06-14' status='unsolved' link_text='See more topics' link_href='https://discourse.getdbt.com/c/show-and-tell/22' />
+          <DiscourseFeed title='Interesting Discussions' category='discussions'  status='unsolved' link_text='See discussions' link_href='https://discourse.getdbt.com/c/discussions/21' />
```
discussions and show and tell don't have Mark as Solved enabled, so the status is redundant here

joellabes reviewed (changes requested):
Yeah this is all basically done from my perspective! Handful of tweaks but then we're good to go

joellabes commented on website/src/components/discourse/index.js:
```diff
@@ -0,0 +1,147 @@
+  title = undefined,
+  link_text = "Ask the Community",
+  link_href = `https://discourse.getdbt.com/new-topic${category ? `?category=${category}` : ''}${tags ? (!category ? `?tags=${tags}` : `&tags=${tags}`) : ''}`,
+  hide_cta = false,
```
Isn't it more standard to have positive flags, so it'd be `show_cta = true`? Non-blocking

JKarlavige committed:
remove in property validation values which wont be used

JKarlavige committed:
default to 90 days for after property for all categories besides 'help' if not set

JKarlavige committed:
setup DiscourseHelpFeed component with predefined props set and set all original props to undefined

JKarlavige committed:
add default props to DiscourseHelpFeed component

JKarlavige committed:
update styles

JKarlavige commented:
> It looks like the meta description is "Description here." Any way to customize that?
> 
> <img alt="Screen Shot 2022-09-19 at 5 18 22 PM" width="648" src="https://user-images.githubusercontent.com/72401230/191118502-d06cec45-7c52-48a1-8c5e-1601aab92326.png">

cc @joellabes, I added title and description meta tags to the /forum page. What should the meta title and description be for the page?

JKarlavige committed:
remove status for show-and-tell and discussions category components

JKarlavige committed:
add like count next to author name and replies if available

JKarlavige committed:
resolve conflict

JKarlavige committed:
move tags prop ahead of link_href in DiscourseHelpFeed

JKarlavige commented:
@joellabes All updates in place and ready for re-review. @lfracalossiFA i'll hold off on requesting a re-review from you until the meta fields are in place.

JKarlavige committed:
keep button text white in dark mode

joellabes commented on website/src/pages/forum.js:
```diff
@@ -1,7 +1,7 @@
-import DiscourseFeed from '@site/src/components/discourse';
+import { DiscourseFeed } from '@site/src/components/discourse';
 
 function Events() {
```
Is this a copy-paste hangover too? And does it matter?

JKarlavige replied:
Since we added a second `DiscourseHelpFeed` component (which i added directly in the original component file), I had to change the way the components were exported. It was originally set to export 1 default component: `export default`, but since the file now has 2 components I changed to named exports `export const <variable-name>`. This allows the imports to be destructured with `import { DiscourseFeed, DiscourseHelpFeed }` as needed

joellabes reviewed (approved):
😻 😻 😻

joellabes commented on website/src/pages/forum.js:
```diff
@@ -0,0 +1,25 @@
+    <Layout>
+      <Head>
+        <title>dbt Discourse</title>
+        <meta name="description" content="Description here" />
```
@JKarlavige How about "Recent interesting discussions from across the dbt Forum" for now

joellabes commented on website/src/components/discourse/index.js:
```diff
@@ -0,0 +1,194 @@
+function TopicWrapper({ topic, children }) {
+  if(topic?.slug) {
+    return (
+      <a href={`https://discourse.getdbt.com/t/${topic.slug}`} title={topic.title} target="_blank">{children}</a>
```
The slug is human-readable, but the actual important part of the link is the topic ID. 

https://discourse.getdbt.com/t/compare-two-models-schemas will redirect to https://discourse.getdbt.com/t/compare-two-models-schemas/3868

https://discourse.getdbt.com/t/fjhsbjhfbshjf/3868 will also get to the right place. 

I don't think this is _likely_ to break since we're using search results from a few seconds ago and the slug probably won't have been changed by a topic rename, but it's worth doing it the proper-er way.

joellabes replied:
(and _THEN_ I promise I'm done with feedback 😅)

JKarlavige replied:
good call, just updated!

JKarlavige committed:
merge current and resolve conflict

JKarlavige committed:
update forum page meta description

JKarlavige committed:
update discourse url path in topicwrapper

JKarlavige committed:
remove console

JKarlavige commented:
All updates in place and re-requested review. Documentation for using this component has been put together and is available here: https://www.notion.so/dbtlabs/Discourse-Component-Documentation-7057e62ee25b4536b84e676ca1c0e892

JKarlavige committed:
add initial tests for DiscourseFeed component

JKarlavige committed:
add tests

JKarlavige committed:
Merge branch 'current' into discourse-integration

JKarlavige committed:
merge current and remove examples from Jinja page

JKarlavige commented:
- Tests added and pass successfully
- Updated to current
- Removed example from Jinja docs page
- This is ready for lift-off 🚀 

![image](https://user-images.githubusercontent.com/13647110/191785008-dae46d73-c04b-463b-8064-722a7b1e9b56.png)

JKarlavige committed:
add forum to Learn dropdown in nav

JKarlavige committed:
nav was reverted, updating to original nav here

JKarlavige committed:
remove false package-lock

JKarlavige commented:
@joellabes note - this will be absent from the navigation until the new top nav is ready to be taken live.
